File: apps/aurity/lib/api/longitudinal-memory.ts
/**
 * Longitudinal Memory API Client
 *
 * Client for the longitudinal memory endpoint that combines
 * chat messages and audio transcriptions chronologically.
 *
 * Card: FI-PHIL-DOC-014 (Memoria Longitudinal)
 * Created: 2025-11-22
 * Updated: 2026-02 - Migrated to centralized api client
 */

import {
  MEMORY_PAGINATION,
  type EventTypeFilter,
} from '@/config/timeline.config';
import { api } from './client';
import { ROUTES } from './routes';
import { createLogger } from '@/lib/internal/logger';

const log = createLogger('LongitudinalMemory');

// API endpoint base — uses ROUTES.timeline as single source of truth
// Backend mounts memory router at /timeline/memory
const MEMORY_BASE = `${ROUTES.timeline}/memory`;
const MEMORY_ENDPOINT = MEMORY_BASE;
const MEMORY_STATS_ENDPOINT = `${MEMORY_BASE}/stats`;

// ============================================================================
// Types (matching backend schemas)
// ============================================================================

export interface MemoryEvent {
  id: string;
  timestamp: number; // Unix timestamp (seconds)
  event_type: 'chat_user' | 'chat_assistant' | 'transcription';
  content: string;
  source: 'chat' | 'audio';

  // Optional metadata
  session_id?: string | null;
  persona?: string | null;
  chunk_number?: number | null;
  duration?: number | null;
  confidence?: number | null;
  language?: string | null;
  stt_provider?: string | null;
}

export interface LongitudinalMemoryResponse {
  events: MemoryEvent[];
  total: number;
  has_more: boolean;
  offset: number;
  limit: number;
  chat_count: number;
  audio_count: number;
  time_range: {
    start: string | null;
    end: string | null;
    start_ts: number | null;
    end_ts: number | null;
  };
}

export interface MemoryStatsResponse {
  total_events: number;
  chat_messages: number;
  audio_transcriptions: number;
  oldest_timestamp: number | null;
  newest_timestamp: number | null;
  unique_sessions: number;
}

export interface MemoryQueryParams {
  doctorId: string;
  offset?: number;
  limit?: number;
  eventType?: EventTypeFilter;
  startTime?: string | null;
  endTime?: string | null;
}

export interface MemorySearchParams {
  doctorId: string;
  query: string;
  offset?: number;
  limit?: number;
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Get longitudinal memory with pagination and filters
 *
 * @param params - Query parameters
 * @returns LongitudinalMemoryResponse
 */
export async function getLongitudinalMemory(
  params: MemoryQueryParams
): Promise<LongitudinalMemoryResponse> {
  const {
    doctorId,
    offset = 0,
    limit = MEMORY_PAGINATION.DEFAULT_LIMIT,
    eventType = 'all',
    startTime,
    endTime,
  } = params;

  const queryParams = new URLSearchParams();
  queryParams.set('doctor_id', doctorId);
  queryParams.set('offset', offset.toString());
  queryParams.set('limit', Math.min(limit, MEMORY_PAGINATION.MAX_LIMIT).toString());
  queryParams.set('event_type', eventType);

  if (startTime) {
    queryParams.set('start_time', startTime);
  }
  if (endTime) {
    queryParams.set('end_time', endTime);
  }

  try {
    const data = await api.get<LongitudinalMemoryResponse>(`${MEMORY_ENDPOINT}?${queryParams}`);
    return data;
  } catch (error) {
    log.error('Fetch failed', { error: String(error) });
    return createEmptyResponse(offset, limit);
  }
}

/**
 * Get memory statistics
 */
export async function getMemoryStats(
  doctorId: string
): Promise<MemoryStatsResponse> {
  try {
    return await api.get<MemoryStatsResponse>(`${MEMORY_STATS_ENDPOINT}?doctor_id=${doctorId}`);
  } catch {
    return createEmptyStats();
  }
}

/**
 * Load more events (for infinite scroll)
 */
export async function loadMoreMemoryEvents(
  params: MemoryQueryParams,
  currentOffset: number
): Promise<LongitudinalMemoryResponse> {
  return getLongitudinalMemory({
    ...params,
    offset: currentOffset,
  });
}

/**
 * Search longitudinal memory by text query
 */
export async function searchMemory(
  params: MemorySearchParams
): Promise<LongitudinalMemoryResponse> {
  const { doctorId, query, offset = 0, limit = MEMORY_PAGINATION.DEFAULT_LIMIT } = params;

  if (!query || query.trim().length === 0) {
    return createEmptyResponse(offset, limit);
  }

  const queryParams = new URLSearchParams();
  queryParams.set('doctor_id', doctorId);
  queryParams.set('query', query.trim());
  queryParams.set('offset', offset.toString());
  queryParams.set('limit', limit.toString());

  try {
    return await api.get<LongitudinalMemoryResponse>(`${MEMORY_BASE}/search?${queryParams}`);
  } catch (error) {
    log.error('Search failed', { error: String(error) });
    return createEmptyResponse(offset, limit);
  }
}

// ============================================================================
// Helper Functions
// ============================================================================

function createEmptyResponse(offset: number, limit: number): LongitudinalMemoryResponse {
  return {
    events: [],
    total: 0,
    has_more: false,
    offset,
    limit,
    chat_count: 0,
    audio_count: 0,
    time_range: {
      start: null,
      end: null,
      start_ts: null,
      end_ts: null,
    },
  };
}

function createEmptyStats(): MemoryStatsResponse {
  return {
    total_events: 0,
    chat_messages: 0,
    audio_transcriptions: 0,
    oldest_timestamp: null,
    newest_timestamp: null,
    unique_sessions: 0,
  };
}

// ============================================================================
// React Query Keys (for cache management)
// ============================================================================

export const memoryQueryKeys = {
  all: ['longitudinal-memory'] as const,
  list: (params: MemoryQueryParams) =>
    [...memoryQueryKeys.all, 'list', params] as const,
  search: (params: MemorySearchParams) =>
    [...memoryQueryKeys.all, 'search', params] as const,
  stats: (doctorId: string) =>
    [...memoryQueryKeys.all, 'stats', doctorId] as const,
};

File: backend/api/routers/memory/public/longitudinal_memory.py
"""Longitudinal Memory API - Memoria Longitudinal (THIN ROUTER).

REFACTORED: All business logic delegated to DIMemoryService.

Single endpoint that combines:
- Chat messages (from conversation memory)
- Audio transcriptions (from sessions HDF5)

All sorted chronologically with pagination and time range filters.

Philosophy: "No existen sesiones. Solo una conversación infinita."

Author: Claude Code
Created: 2025-11-22
Refactored: 2026-01-28 (Phase 2.3 - DI pattern)
Card: Backend Refactor Phase 2.3 - Service Refactoring
"""

from __future__ import annotations

from enum import Enum
from typing import Literal

from backend.services.memory.dependencies import get_memory_service
from backend.services.memory.services.di_memory_service import DIMemoryService
from backend.utils.common.logging.logger import get_logger
from fastapi import APIRouter, Depends, Query, status
from pydantic import BaseModel, Field

logger = get_logger(__name__)

router = APIRouter(prefix="/timeline/memory")


# ============================================================================
# Configuration (no hardcoding)
# ============================================================================


class TimelineConfig:
    """Centralized configuration for timeline limits."""

    DEFAULT_LIMIT: int = 50
    MAX_LIMIT: int = 200
    DEFAULT_WINDOW_HOURS: int = 24 * 7  # 1 week default
    MAX_WINDOW_DAYS: int = 365  # 1 year max


# ============================================================================
# Types
# ============================================================================


class EventType(str, Enum):
    """Event types for filtering."""

    ALL = "all"
    CHAT = "chat"
    AUDIO = "audio"


class MemoryEvent(BaseModel):
    """Single event in the unified timeline."""

    id: str = Field(..., description="Unique event ID")
    timestamp: int = Field(..., description="Unix timestamp (seconds)")
    event_type: Literal["chat_user", "chat_assistant", "transcription"] = Field(
        ..., description="Type of event"
    )
    content: str = Field(..., description="Event content/text")
    source: Literal["chat", "audio"] = Field(..., description="Data source")

    # Optional metadata (all default to None)
    session_id: str | None = Field(default=None, description="Session ID if applicable")
    persona: str | None = Field(default=None, description="AI persona (for assistant messages)")
    chunk_number: int | None = Field(default=None, description="Audio chunk number")
    duration: float | None = Field(default=None, description="Audio duration in seconds")
    confidence: float | None = Field(default=None, description="Transcription confidence")
    language: str | None = Field(default=None, description="Detected language")
    stt_provider: str | None = Field(default=None, description="STT provider used")


class LongitudinalMemoryResponse(BaseModel):
    """Response for longitudinal memory endpoint."""

    events: list[MemoryEvent] = Field(default_factory=list)
    total: int = Field(..., description="Total events matching filters")
    has_more: bool = Field(..., description="More events available")
    offset: int = Field(..., description="Current offset")
    limit: int = Field(..., description="Items per page")

    # Metadata
    chat_count: int = Field(0, description="Number of chat events in response")
    audio_count: int = Field(0, description="Number of audio events in response")
    time_range: dict = Field(default_factory=dict, description="Applied time range")


class MemoryStatsResponse(BaseModel):
    """Statistics for the unified timeline."""

    total_events: int = Field(0)
    chat_messages: int = Field(0)
    audio_transcriptions: int = Field(0)
    oldest_timestamp: int | None = Field(None)
    newest_timestamp: int | None = Field(None)
    unique_sessions: int = Field(0)


# ============================================================================
# Note: Helper functions removed - now in DIMemoryService
# - _parse_timestamp() → service method
# - _get_chat_events() → service method
# - _get_audio_events() → service method
# ============================================================================


# ============================================================================
# Endpoints
# ============================================================================


@router.get(
    "",
    response_model=LongitudinalMemoryResponse,
    status_code=status.HTTP_200_OK,
)
async def get_longitudinal_memory(
    doctor_id: str = Query(..., description="Doctor ID (Auth0 user.sub)"),
    offset: int = Query(0, ge=0, description="Number of events to skip"),
    limit: int = Query(
        TimelineConfig.DEFAULT_LIMIT,
        ge=1,
        le=TimelineConfig.MAX_LIMIT,
        description="Events per page",
    ),
    event_type: EventType = Query(EventType.ALL, description="Filter by event type"),
    start_time: str | None = Query(
        None,
        description="Start time (ISO 8601 or Unix timestamp)",
    ),
    end_time: str | None = Query(
        None,
        description="End time (ISO 8601 or Unix timestamp)",
    ),
    service: DIMemoryService = Depends(get_memory_service),
) -> LongitudinalMemoryResponse:
    """Get longitudinal memory combining chat messages and audio transcriptions (THIN ROUTER).

    All events are sorted chronologically (newest first) and can be filtered
    by type and time range.

    Philosophy: "No existen sesiones. Solo una conversación infinita."

    Args:
        doctor_id: Doctor identifier (Auth0 user.sub)
        offset: Number of events to skip for pagination
        limit: Maximum events to return (max 200)
        event_type: Filter by "all", "chat", or "audio"
        start_time: Optional start of time range
        end_time: Optional end of time range
        service: Memory service (injected by FastAPI Depends)

    Returns:
        LongitudinalMemoryResponse with merged, sorted events
    """
    # Delegate to service (all business logic)
    result = await service.get_longitudinal_memory(
        doctor_id=doctor_id,
        offset=offset,
        limit=limit,
        event_type=event_type.value,
        start_time=start_time,
        end_time=end_time,
    )

    # Map service result to API response
    return LongitudinalMemoryResponse(**result)


@router.get(
    "/search",
    response_model=LongitudinalMemoryResponse,
    status_code=status.HTTP_200_OK,
)
async def search_memory(
    doctor_id: str = Query(..., description="Doctor ID (Auth0 user.sub)"),
    query: str = Query(..., min_length=1, max_length=500, description="Search query"),
    limit: int = Query(50, ge=1, le=TimelineConfig.MAX_LIMIT),
    offset: int = Query(0, ge=0),
    service: DIMemoryService = Depends(get_memory_service),
) -> LongitudinalMemoryResponse:
    """Search longitudinal memory by text query (THIN ROUTER).

    Uses case-insensitive substring matching across content.
    Searches both chat messages and audio transcriptions.

    Returns results sorted by timestamp (newest first).
    Future enhancement: Semantic search with embeddings.

    Args:
        doctor_id: Doctor identifier (Auth0 user.sub)
        query: Search text (1-500 characters)
        limit: Maximum events to return (max 200)
        offset: Number of events to skip for pagination
        service: Memory service (injected by FastAPI Depends)

    Returns:
        LongitudinalMemoryResponse with search results
    """
    # Delegate to service (all business logic)
    result = await service.search_memory(
        doctor_id=doctor_id,
        query=query,
        limit=limit,
        offset=offset,
    )

    # Map service result to API response
    return LongitudinalMemoryResponse(**result)


@router.get(
    "/stats",
    response_model=MemoryStatsResponse,
    status_code=status.HTTP_200_OK,
)
async def get_memory_stats(
    doctor_id: str = Query(..., description="Doctor ID (Auth0 user.sub)"),
    service: DIMemoryService = Depends(get_memory_service),
) -> MemoryStatsResponse:
    """Get statistics for the longitudinal memory (THIN ROUTER).

    Provides counts and metadata without fetching full events.
    Useful for UI indicators and pagination info.

    Args:
        doctor_id: Doctor identifier (Auth0 user.sub)
        service: Memory service (injected by FastAPI Depends)

    Returns:
        MemoryStatsResponse with aggregate statistics
    """
    # Delegate to service (all business logic)
    result = await service.get_stats(doctor_id=doctor_id)

    # Map service result to API response
    return MemoryStatsResponse(**result)
